python/ray/serve/_private/deployment_state.py

```diff
@@ -1158,6 +1180,23 @@ def _set_target_state(self, target_info: DeploymentInfo) -> None:
         target_state = DeploymentTargetState.from_deployment_info(target_info)
         self._save_checkpoint_func(writeahead_checkpoints={self._name: target_state})
 
+        if self._target_state.version == target_state.version:
```

Comment:
confused by this block -- shouldn't the equality check only return `True` if both code_version and config match? therefore the conditions below would never pass?

Reply:
This is what I use for `__hash__` in `DeploymentVersion`:
<img width="567" alt="image" src="https://user-images.githubusercontent.com/15851518/233727121-fdbe6594-658a-419a-b3ff-f23fee444edc.png">

So basically `num_replicas` and `autoscaling_config` (the only two lightweight options) are excluded when calculating version.

Reply:
Ah, I get it now. I find this pretty confusing -- intuitively I'd expect the equality check to check _all_ options and then maybe we'd have a separate method for checking partial equality like this

Reply:
Hm yeah I can see how that might be confusing ... 
I think version is used specifically for detecting "old replicas with outdated versions" that need updating (see [code pointer](https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/deployment_state.py#L1240)). So when the target state is updated with new `num_replicas` and `autoscaling_config` values, that is automatically picked up by deployment state manager, but has nothing to do with updating old replicas.

Perhaps a better way to deal with this is rename DeploymentVersion to ReplicaVersion lol. Maybe we can also move options like `max_concurrent_queries` into `ReplicaConfig` instead of `DeploymentConfig`.

Reply:
that makes a lot of sense to me... `ReplicaConfig` is a subset of `DeploymentConfig` and contains all of the things that get passed to each replica

if it's a big change do it in a follow up?